assembly/src/assembler/mod.rs

```diff
@@ -62,18 +60,31 @@ pub struct Assembler {
 // ------------------------------------------------------------------------------------------------
 /// Constructors
 impl Assembler {
-    /// Start building an [Assembler]
-    pub fn new() -> Self {
-        Self::default()
-    }
+    /// Start building an [`Assembler`] with a kernel defined by the provided [CompiledLibrary].
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The library does not contain a kernel module.
+    /// - The library contains non-kernel modules.
+    pub fn with_kernel(kernel_lib: CompiledLibrary) -> Result<Self, AssemblerError> {
```

Comment:
This feels like something we can solve with a dedicated type like `Program`, i.e. `Kernel` wraps a `CompiledLibrary`, and by construction meets all of the kernel library criteria. Then this function can be changed to only accept a `Kernel`, rather than any `CompiledLibrary` (and when assembling a kernel, you get a `Kernel` as output).

All validation of a `Kernel` can be done during assembly (and deserialization when appropriate), and you get to elide this type of validation entirely everywhere else.

Reply:
Yep - introducing `KernelLibrary` removed the need for this error checking.

I kept `Kernel` and `KernelLibrary` separate for now (mostly because the verifier does not need to know anything besides the procedure MAST roots. But it does feel like there could be some way to improve the structure. Specifically, right now we have 3 structs involved in this:

1. `Kernel` which holds only procedure digests.
2. `KernelLibrary` which holds the underlying MAST for the kernel.
3. `ModuleInfo` which holds digests + procedure paths (this is used by non-kernel modules as well).

There is some overlap between what these structs hold, and I feel like there should be a way to consolidate these into 2 structs somehow.